File: Python/solutions/20-num-repeats.py
# Write a method that takes in a string and returns the number of
# letters that appear more than once in the string. You may assume
# the string contains only lowercase letters. Count the number of
# letters that repeat, not the number of times they repeat in the
# string.
#
# Difficulty: hard.

def num_repeats(string):
    tally = {}

    for letter in string:
        if not letter in tally.keys():
            tally[letter] = 0
        tally[letter] += 1

    repeat_count = 0
    for key, value in tally.items():
        if value > 1:
            repeat_count += 1

    return repeat_count

# These are tests to check that your code is working. After writing
# your solution, they should all print True.

print('num_repeats("abdbc") == 1: ' + str(num_repeats('abdbc') == 1))
print('num_repeats("aaa") == 1: ' + str(num_repeats('aaa') == 1))
print('num_repeats("abab") == 2: ' + str(num_repeats('abab') == 2))
print('num_repeats("cadac") == 2: ' + str(num_repeats('cadac') == 2))
print('num_repeats("abcde") == 0: ' + str(num_repeats('abcde') == 0))

File: Ruby/solutions (from jayqui)/20-num-repeats.rb
# Write a method that takes in a string and returns the number of
# letters that appear more than once in the string. You may assume
# the string contains only lowercase letters. Count the number of
# letters that repeat, not the number of times they repeat in the
# string.
#
# Difficulty: hard.

def num_repeats(string)
  tallies = {}

  (0..string.length - 1).each do |idx|
    letter = string[idx]
    tallies[letter] ||= 0
    tallies[letter] += 1
  end

  count_tallies_greater_than_two(tallies)
end

def count_tallies_greater_than_two(tally_list)
  result = 0

  tally_list.each do |key, value|
    if value >= 2
      result += 1
    end
  end

  result

  ## ^ The above is roughly equivalent to:
  # tally_list.select {|key, value| key if value >= 2}.count
end

# These are tests to check that your code is working. After writing
# your solution, they should all print true.

puts('num_repeats("abdbc") == 1: ' + (num_repeats('abdbc') == 1).to_s)
# one character is repeated
puts('num_repeats("aaa") == 1: ' + (num_repeats('aaa') == 1).to_s)
puts('num_repeats("abab") == 2: ' + (num_repeats('abab') == 2).to_s)
puts('num_repeats("cadac") == 2: ' + (num_repeats('cadac') == 2).to_s)
puts('num_repeats("abcde") == 0: ' + (num_repeats('abcde') == 0).to_s)
